core/vm/eips.go

```diff
@@ -1055,13 +1060,20 @@ func opExtCall(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]
 	if interpreter.readOnly && !value.IsZero() {
 		return nil, ErrWriteProtection
 	}
-	if !value.IsZero() {
-		gas += params.CallStipend
-	}
 
-	ret, returnGas, err := interpreter.evm.Call(scope.Contract, toAddr, args, gas, &value)
+	var (
+		ret       []byte
+		returnGas uint64
+		err       error
+	)
+	if interpreter.evm.callGasTemp == 0 {
+		// zero temp call gas indicates a min retained gas error
+		ret, returnGas, err = nil, 0, ErrExecutionReverted
+	} else {
+		ret, returnGas, err = interpreter.evm.Call(scope.Contract, toAddr, args, gas, &value)
+	}
 
-	if err == ErrExecutionReverted {
+	if err == ErrExecutionReverted || err == ErrInsufficientBalance || err == ErrDepth {
```

Comment:
Better to use `errors.Is`. But ... do we now have evm instrospection into exactly what type of error that occurs? that sounds crazy

Reply:
```
1 if the call has reverted (also can be pushed earlier in a light failure scenario).
2 if the call has failed.
```
What does failed mean? Doesn't `err` here always signify that it has reverted?

Reply:
Generally exit code 1 means something about the VM prevented the call (Insufficient balance, stack to deep, out of gas, other exceptional halts) and exit code 2 means the called code explicitly caused the failure, typically by invoking a REVERT opcode.

Reply:
If there's a better way to filter this (such as testing for return code 2 and 1 becomes the fall through) let me know, I'm not up to date on all of the geth execution error code names.  This also would be a good indicator that we need more reference tests attempting to trigger all the other error cases that could be returned here.

Reply:
What about precompiles? Our precompiles might return errors that are not caught by this and will return 2, which seems like the wrong approach to me. If this should only happen because of REVERT, than I think we need to explicitly test for REVERT

Reply:
> Generally exit code 1 means something about the VM prevented the call (Insufficient balance, stack to deep, out of gas, other exceptional halts) and exit code 2 means the called code explicitly caused the failure, typically by invoking a REVERT opcode.

Wait isn't it the other way around?

Reply:
> What about precompiles?

I've noticed in the past, that sometimes EIP-authors proposing new precompiles have been used formulations like "throw an error if xx", or "if input is bad, treat as OOG". It's worked out ok, because we haven't differentiated across errors: only "eat all gas or not?".

Reply:
I think all precompiles would be return code 1 failures, unless it is explicitly attempting to simulate a REVERT operation.